modules/javafx.graphics/src/main/native-prism-mtl/MetalTexture.m

```diff
@@ -0,0 +1,459 @@
+    id<MTLBuffer> pixelMTLBuf = nil;
+    int offset = [[context getDataRingBuffer] reserveBytes:length];
+    if (offset < 0) {
+        pixelMTLBuf = [context getTransientBufferWithLength:length];
+        offset = 0;
+    } else {
+        pixelMTLBuf = [[context getDataRingBuffer] getBuffer];
+    }
+
+    [bufferOffsetDict setObject:pixelMTLBuf forKey:@(offset)];
+    return bufferOffsetDict;
+}
+
+static NSMutableDictionary *copyPixelDataToRingBuffer(MetalContext* context, void* pixels,
+    int srcx, int srcy, int w, int h, int scanStride, MTLPixelFormat pixelFormat)
+{
+    unsigned int pixelSize = getPixelSize(pixelFormat);
+    unsigned int length = pixelSize * w * h;
+    NSMutableDictionary<NSNumber *, id<MTLBuffer>> *bufferOffsetDict = getBufferAndOffset(context, length);
+    NSNumber *offset = [[bufferOffsetDict allKeys] firstObject];
+    id<MTLBuffer> dstBuf = [[bufferOffsetDict allValues] firstObject];
+
+    void *dstBufOffset = dstBuf.contents + [offset intValue];
+    unsigned int rowLength = pixelSize * w;
+    void *pixelsSrcOffset = pixels + srcy * scanStride + srcx * pixelSize;
+
+    for (int i = 0; i < h; i++) {
+        memcpy(dstBufOffset + (rowLength * i), pixelsSrcOffset + (scanStride * i), rowLength);
+    }
+
+    return bufferOffsetDict;
+}
+
+
+// ** JNI METHODS **
+
+/*
+ * Class:     com_sun_prism_mtl_MTLTexture
+ * Method:    nUpdate
+ * Signature: (JJLjava/nio/ByteBuffer;[BIIIIIII)J
+ */
+JNIEXPORT jlong JNICALL Java_com_sun_prism_mtl_MTLTexture_nUpdate
+    (JNIEnv *env, jclass jClass, jlong ctx, jlong nTexturePtr, jobject buf,
+    jbyteArray pixData, jint dstx, jint dsty, jint srcx, jint srcy,
+    jint w, jint h, jint scanStride)
+{
+    MetalContext* context = (MetalContext*)jlong_to_ptr(ctx);
+    MetalTexture* mtlTex  = (MetalTexture*)jlong_to_ptr(nTexturePtr);
+
+    jint length = pixData?
+        (*env)->GetArrayLength(env, pixData) :
+        (jint)((*env)->GetDirectBufferCapacity(env, buf));
+    length *= sizeof(jbyte);
+
+    jbyte* pixels = (jbyte*)((pixData != NULL) ?
+        (*env)->GetPrimitiveArrayCritical(env, pixData, NULL) :
+        (*env)->GetDirectBufferAddress(env, buf));
+
+    id<MTLTexture> tex = [mtlTex getTexture];
+    MTLPixelFormat pixelFormat = [mtlTex getPixelFormat];
+
+    NSMutableDictionary* bufferOffsetDict = copyPixelDataToRingBuffer(context, pixels, srcx, srcy,
+                                                                    w, h, scanStride, pixelFormat);
+    int offset = [[[bufferOffsetDict allKeys] firstObject] intValue];
+    id<MTLBuffer> pixelMTLBuf = [[bufferOffsetDict allValues] firstObject];
+
+    if (pixData != NULL) {
+        (*env)->ReleasePrimitiveArrayCritical(env, pixData, pixels, 0);
+    }
+
+    [context endCurrentRenderEncoder];
+    id<MTLCommandBuffer> commandBuffer = [context getCurrentCommandBuffer];
+    @autoreleasepool {
+        id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer blitCommandEncoder];
```

Comment:
Similarly to `MetalContext` comment, I think we should separate JNI entry point code and argument parsing/converting from actual Metal-reaching implementation. I would also move this to a `MetalTexture` method.